src/Framework/ReuseableStringBuilder.cs

```diff
@@ -0,0 +1,346 @@
+                {
+                    // Allocation of arrays is optimized in byte[bytes] => bytes = 2^n.
+                    // StringBuilder allocates chars[capacity] and each char is 2 bytes so lets have capacity brackets computed as `bytes/2` 
+                    brackets.Add(bytes/2); 
+                    bytes <<= 1;
+                }
+                Debug.Assert((bytes >> 1) == MaxBuilderSizeBytes, "MaxBuilderSizeBytes has to be 2^n (power of 2)");
+
+                s_capacityBrackets = brackets;
+            }
+
+            /// <summary>
+            /// The shared builder.
+            /// </summary>
+            private static StringBuilder s_sharedBuilder;
+
+#if DEBUG && ASSERT_BALANCE
+            /// <summary>
+            /// Balance between calling Get and Release.
+            /// Shall be always 0 as Get and 1 at Release.
+            /// </summary>
+            private static int s_getVsReleaseBalance;
+#endif
+
+            /// <summary>
+            /// Obtains a string builder which may or may not already
+            /// have been used. 
+            /// Never returns null.
+            /// </summary>
+            internal static StringBuilder Get(int capacity)
+            {
+#if DEBUG && ASSERT_BALANCE
+                int balance = Interlocked.Increment(ref s_getVsReleaseBalance);
+                Debug.Assert(balance == 1, "Unbalanced Get vs Release. Either forgotten Release or used from multiple threads concurrently.");
+#endif
+
+                var returned = Interlocked.Exchange(ref s_sharedBuilder, null);
+
+                if (returned == null)
+                {
+                    // Currently loaned out so return a new one with capacity in given bracket.
+                    // If user wants bigger capacity that maximum capacity respect it.
+                    returned = new StringBuilder(SelectBracketedCapacity(capacity));
+#if DEBUG
+                    MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: returned.GetHashCode(), newCapacity:capacity, oldCapacity:0, type:"miss");
+#endif
+                }
+                else if (returned.Capacity < capacity)
+                {
+                    // It's essential we guarantee the capacity because this
+                    // may be used as a buffer to a PInvoke call.
+                    int newCapacity = SelectBracketedCapacity(capacity);
+#if DEBUG
+                    MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: returned.GetHashCode(), newCapacity: newCapacity, oldCapacity: returned.Capacity, type: "miss-need-bigger");
+#endif
+                    // Let the current StringBuilder be collected and create new with bracketed capacity. This way it allocates only char[newCapacity]
+                    //   otherwise it would allocate char[new_capacity_of_last_chunk] (in set_Capacity) and char[newCapacity] (in Clear).
+                    returned = new StringBuilder(SelectBracketedCapacity(newCapacity));
```

Comment:
Was this supposed to be:

```suggestion
                    returned = new StringBuilder(newCapacity);
```

If not, why call `SelectBracketedCapacity` a second time, passing its own output to itself?

Reply:
You are right. good catch.